src/main/java/net/fabricmc/loader/impl/launch/FabricLauncherBase.java

```diff
@@ -63,152 +65,181 @@ public MappingConfiguration getMappingConfiguration() {
+
 					Log.debug(LogCategory.GAME_REMAP, "Appending '%s' to remapper classpath", path);
-					remapper.readClassPath(path);
+					remapper.readClassPathAsync(path);
 				}
+			} catch (URISyntaxException e) {
+				throw new RuntimeException("Failed to convert '" + url + "' to path!", e);
+			}
+		}
 
-				remapper.readInputs(jarFile);
-				remapper.apply(outputConsumer);
-			} finally {
-				remapper.finish();
+		List<OutputConsumerPath> outputConsumers = new ArrayList<>(inputFiles.size());
+		List<InputTag> inputTags = new ArrayList<>(inputFiles.size());
+
+		try {
+			for (int i = 0; i < inputFiles.size(); i++) {
+				Path inputFile = inputFiles.get(i);
+				Path tmpFile = tmpFiles.get(i);
+
+				InputTag inputTag = remapper.createInputTag();
+				OutputConsumerPath outputConsumer = new OutputConsumerPath.Builder(tmpFile)
+						// force jar despite the .tmp extension
+						.assumeArchive(true)
+						.build();
+
+				outputConsumers.add(outputConsumer);
+				inputTags.add(inputTag);
+
+				outputConsumer.addNonClassFiles(inputFile, NonClassCopyMode.FIX_META_INF, remapper);
+				remapper.readInputsAsync(inputTag, inputFile);
 			}
 
-			// Minecraft doesn't tend to check if a ZipFileSystem is already present,
-			// so we clean up here.
+			for (int i = 0; i < inputFiles.size(); i++) {
+				remapper.apply(outputConsumers.get(i), inputTags.get(i));
+			}
+		} finally {
+			for (OutputConsumerPath outputConsumer : outputConsumers) {
+				outputConsumer.close();
+			}
 
-			depPaths.add(deobfJarFileTmp);
+			remapper.finish();
+		}
 
-			for (Path p : depPaths) {
-				try {
-					p.getFileSystem().close();
-				} catch (Exception e) {
-					// pass
-				}
+		// Minecraft doesn't tend to check if a ZipFileSystem is already present,
+		// so we clean up here.
 
-				try {
-					FileSystems.getFileSystem(new URI("jar:" + p.toUri())).close();
-				} catch (Exception e) {
-					// pass
-				}
+		depPaths.addAll(tmpFiles);
+
+		for (Path p : depPaths) {
+			try {
+				p.getFileSystem().close();
+			} catch (Exception e) {
+				// pass
+			}
+
+			try {
+				FileSystems.getFileSystem(new URI("jar:" + p.toUri())).close();
+			} catch (Exception e) {
+				// pass
 			}
+		}
+
+		List<Path> missing = new ArrayList<>();
+
+		for (int i = 0; i < inputFiles.size(); i++) {
+			Path inputFile = inputFiles.get(i);
+			Path tmpFile = tmpFiles.get(i);
+			Path outputFile = outputFiles.get(i);
 
-			try (JarFile jar = new JarFile(deobfJarFileTmp.toFile())) {
+			boolean found;
+
+			try (JarFile jar = new JarFile(tmpFile.toFile())) {
 				found = jar.stream().anyMatch((e) -> e.getName().endsWith(".class"));
 			}
 
 			if (!found) {
-				Log.error(LogCategory.GAME_REMAP, "Generated deobfuscated JAR contains no classes! Trying again...");
-				Files.delete(deobfJarFileTmp);
+				missing.add(inputFile);
+				Files.delete(tmpFile);
 			} else {
-				Files.move(deobfJarFileTmp, deobfJarFile);
+				Files.move(tmpFile, outputFile);
 			}
-		} while (!found);
```

Comment:
Previously it would keep trying if a remap fails, might be worth re-attempting once if it was just a fluke rather than immediately giving up

Reply:
I want to see if there's actually a problem this solves, something is particularly wrong otherwise.